test/manipulation.test.js

```diff
@@ -1686,7 +1700,73 @@ describe('manipulation', function() {
     });
 
     // TODO: implement destroy with filtered set
-    it('should destroy filtered set of records');
+    it.skip('should destroy filtered set of records');
+
+    it('should not delete using condition from filter', function(done) {
+      Person.destroyAll({where: {name: 'John'}}, function(err, info) {
+        should.exist(err);
+        err.should.match(/Filter object detected/);
+        Person.find({fields: 'name'}, function(err, users) {
+          should.not.exist(err);
+          users.should.containDeep([
+            {name: 'Jane'},
+            {name: 'Mary'},
+            {name: 'John'},
+          ]);
+          done();
+        });
+      });
+    });
+
+    it('should not delete all for bad and condition', function(done) {
+      Person.destroyAll({and: [{name: 'Jane'}, {where: {name: 'John'}}]}, function(err, info) {
+        should.not.exist(err);
+        Person.find({fields: 'name'}, function(err, users) {
+          should.not.exist(err);
+          let userObjects = [];
+          users.forEach(user => userObjects.push(user.toObject()));
+          userObjects.should.containDeep([
+            {name: 'John'},
+            {name: 'Jane'},
+            {name: 'Mary'},
+          ]);
+          done();
+        });
+      });
+    });
+
+    it('should delete all for undefined object', function(done) {
+      Person.destroyAll(undefined, function(err, info) {
+        should.not.exist(err);
+        Person.find({fields: 'name'}, function(err, users) {
+          should.not.exist(err);
+          users.should.containDeep([]);
```

Comment:
Isn't this assertion always true as long as `users` is any array? I think you need to call `users.should.deepEqual` instead.